src/Microsoft.AspNetCore.SignalR.Common/Protocol/HandshakeProtocol.cs

```diff
@@ -17,28 +19,39 @@ public static class HandshakeProtocol
     {
         private const string ProtocolPropertyName = "protocol";
         private const string ProtocolVersionPropertyName = "version";
+        private const string MinorVersionName = "minorVersion";
         private const string ErrorPropertyName = "error";
         private const string TypePropertyName = "type";
 
-        /// <summary>
-        /// The serialized representation of a success handshake.
-        /// </summary>
-        public static ReadOnlyMemory<byte> SuccessHandshakeData;
+        [Obsolete()]
+        public static ReadOnlyMemory<byte> SuccessHandshakeData = new byte[] { 123, 34, 109, 105, 110, 111, 114, 86, 101, 114, 115, 105, 111, 110, 34, 58, 48, 125, 30 }; 
+
+
+        private static Dictionary<IHubProtocol, ReadOnlyMemory<byte>> _messageCache = new Dictionary<IHubProtocol, ReadOnlyMemory<byte>>();
 
-        static HandshakeProtocol()
+        public static ReadOnlySpan<byte> GetCachedSuccessMessage(IHubProtocol protocol)
         {
+            if (_messageCache.TryGetValue(protocol, out var message))
+            {
+                return message.Span;
+            }
+
             var memoryBufferWriter = MemoryBufferWriter.Get();
             try
             {
-                WriteResponseMessage(HandshakeResponseMessage.Empty, memoryBufferWriter);
-                SuccessHandshakeData = memoryBufferWriter.ToArray();
+                WriteResponseMessage(new HandshakeResponseMessage(protocol), memoryBufferWriter);
+                _messageCache.Add(protocol, memoryBufferWriter.ToArray());
             }
             finally
             {
                 MemoryBufferWriter.Return(memoryBufferWriter);
             }
+
+            _messageCache.TryGetValue(protocol, out var result);
```

Comment:
Return a span from `memoryBufferWriter.ToArray()` up above. Avoids adding the data to the cache then getting it again for no reason.

Reply:
This would look like `return (ReadOnlySpan<byte>)memoryBufferWriter.ToArray();`

I kind of like the previous way, because the symmetry between the two returns makes the function easier to read, at least to me. This is run 2-3 times total per server, so perf concerns are minimal.